_posts/2023-04-19-more-optimizations.md

```diff
@@ -0,0 +1,1300 @@
+
+```
+    {test,is_integer,{f,19},[{x,0}]}.
+```
+
+It is followed by the guard test `X =< 0`:
+
+```
+    {test,is_ge,{f,19},[{tr,{x,0},{t_integer,any}},{integer,0}]}.
+```
+
+As a result of the `is_integer/1` test, it is known that `{x,0}`
+is an integer.
+
+The third instruction corresponds to `X < 16#400`, which the compiler rewrote
+to `16#3FF =< X` (`1023 =< X`):
+
+```
+    {test,is_ge,{f,19},[{integer,1023},{tr,{x,0},{t_integer,{0,'+inf'}}}]}.
+```
+
+In the type for the `{x,0}` register there is something new for OTP
+26. It says that the range is 0 through `'+inf'`, that is, from 0 up
+to positive infinity. From that information, the compiler can infer
+that if this instruction succeeds, the type for `{x,0}` is
+`t_integer,{0,1023}}`.
+
+
+### Combining guard tests
+
+In OTP 25, the JIT would emit native code for each BEAM instruction
+in the guard individually. When translated individually, the three guards
+tests for one of the variables each require 11 native instructions, or 33
+instructions for all three.
+
+By having the BEAM loader combining the three guard tests into a
+single `is_int_range` instruction, the JIT is capable of doing a much
+better job and emit a mere 6 native instructions.
+
+How is that possible?
+
+As individual BEAM instructions, each guard test needs 5 instructions
+to fetch the value from `{x,0}` and test that the value is a small
+integer. As a combined instruction, that only needs to be done once.
+Other parts of the guard tests also become redundant in the combined
+instruction and can be omitted. For example, the `is_integer/1` type
+test will also succeed if its argument is a **bignum** (an integer
+that does not fit in a machine word). Clearly, a bignum will fall well
+outside the range 0 through 1023, so if the argument is not a small
+integer, the combined guard test will fail immediately.
+
+With those and some other simplifications, we end up with the following
+native instructions:
+
+```nasm
+# is_int_in_range_fScc
+    mov rax, qword ptr [rbx]
+    sub rax, 15
+    test al, 15
+    short jne label_19
+    cmp rax, 16368
+    short ja label_19
+```
+
+The first instruction fetches the value of `{x,0}` to the CPU
+register `rax`:
+
+```nasm
+    mov rax, qword ptr [rbx]
+```
+
+The next instruction subtracts the [tagged value][tags] of the lower
+bound of the range. Since the lower bound of the range is 0 and the
+tag for small integers is 15, the value that is subtracted is
+`16 * 0 + 15` or simply 15. If the lower bound would have been 1, the
+value to be subtracted would have been `16 * 1 + 15` or 31:
+
+```nasm
+    sub rax, 15
+```
+
+The subtraction achieves two aims at once. Firstly, it simplifies the
+tag test in the next two instructions because if the value of of
+`{x,0}` is a small integer, the 4 least significants bits will now be
+zero:
+
+```nasm
+    test al, 15
+    short jne label_19
+```
+
+The `test al, 15` instruction does a bitwise AND operation of the
```

Comment:
Stupid question: I had the notion that a 'test' instruction subtracted without result..., not bitwise AND?

Reply:
Oh, wait...was that 'cmp', not 'test'?  I am probably out in the woods here

Reply:
Yes, [test](https://www.felixcloutier.com/x86/test) does a bitwise AND, [cmp](https://www.felixcloutier.com/x86/cmp) subtracts.